x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/TopNOperator.java

```diff
@@ -273,73 +496,59 @@ public String describe() {
-                @Override
-                protected boolean lessThan(Row a, Row b) {
-                    return TopNOperator.compareTo(sortOrders, a, b) < 0;
-                }
-
-                @Override
-                public String toString() {
+                } else {
                     return "count = " + size() + "/" + topCount + ", sortOrders = " + sortOrders;
                 }
-            };
-        }
-    }
-
-    private static int compareTo(List<SortOrder> orders, Row a, Row b) {
-        for (SortOrder order : orders) {
-            int cmp = comparePositions(order.asc, order.nullsFirst, a, b, order.channel);
-            if (cmp != 0) {
-                return cmp;
             }
-        }
-        return 0;
+        };
     }
 
-    static int comparePositions(boolean asc, boolean nullsFirst, Row b1, Row b2, int position) {
-        boolean firstIsNull = b1.isNull(position);
-        boolean secondIsNull = b2.isNull(position);
-        if (firstIsNull || secondIsNull) {
-            return Boolean.compare(firstIsNull, secondIsNull) * (nullsFirst ? 1 : -1);
+    static int compareRows(Row r1, Row r2) {
+        // This is simliar to r1.orderByCompositeKey.compareTo(r2.orderByCompositeKey) but stopping somewhere in the middle so that
+        // we check the byte that mismatched
+        BytesRef br1 = r1.orderByCompositeKey.get();
+        BytesRef br2 = r2.orderByCompositeKey.get();
+        int mismatchedByteIndex = Arrays.mismatch(
+            br1.bytes,
+            br1.offset,
+            br1.offset + br1.length,
+            br2.bytes,
+            br2.offset,
+            br2.offset + br2.length
+        );
+        if (mismatchedByteIndex < 0) {
+            // the two rows are equal
+            return 0;
         }
-        if (b1.idToType[position] != b2.idToType[position]) {
-            throw new IllegalStateException(
-                "Blocks have incompatible element types: " + b1.idToType[position] + " != " + b2.idToType[position]
+        int length = Math.min(br1.length, br2.length);
+        // one value is the prefix of the other
+        if (mismatchedByteIndex == length) {
+            // the value with the greater length is considered greater than the other
+            if (length == br1.length) {// first row is less than the second row
+                return r2.orderByCompositeKeyAscending.get(length) ? 1 : -1;
+            } else {// second row is less than the first row
+                return r1.orderByCompositeKeyAscending.get(length) ? -1 : 1;
+            }
+        } else {
+            // compare the byte that mismatched accounting for that respective byte asc/desc ordering
+            int c = Byte.compareUnsigned(
+                r1.orderByCompositeKey.bytes()[br1.offset + mismatchedByteIndex],
+                r2.orderByCompositeKey.bytes()[br2.offset + mismatchedByteIndex]
             );
+            return r1.orderByCompositeKeyAscending.get(mismatchedByteIndex) ? -c : c;
```

Comment:
It'd be amazing if we could get away with a single comparison of the whole array, but that'll be fiddly and can wait. This works.